### PR TITLE
fix: fix the case that rollout stuck when there are failed to apply binding

### PR DESCRIPTION
**Pull request by ryanzhang-oss**
### Description of your changes

Rollout can get stuck when there are failed to apply bindings since there are already max unavailable bindings. The fix is to update the failed bindings first since it won't make more bindings unavailable since they are already unavailable anyway.

Fixes #

I have:

- [x ] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

UT/IT/E2E

### CODE REVIEW

File: pkg/controllers/rollout/controller.go [hunk 1]
```diff
@@ -329,6 +352,11 @@ func pickBindingsToRoll(allBindings []*fleetv1beta1.ClusterResourceBinding, late
 	klog.V(2).InfoS("Calculated the max number of bindings to remove", "clusterResourcePlacement", klog.KObj(crp),
 		"maxUnavailableNumber", maxUnavailableNumber, "minAvailableNumber", minAvailableNumber,
 		"lowerBoundAvailableBindings", lowerBoundAvailableNumber, "maxNumberOfBindingsToRemove", maxNumberToRemove)
+
+	// we can still update the bindings that are failed to apply already regardless of the maxNumberToRemove
+	for i := 0; i < maxUnavailableNumber && i < len(applyFailedUpdateCandidates); i++ {
```
Comment (zhiying-lin) [inline pkg/controllers/rollout/controller.go:357]: we don't care about the maxUnavailableNumber, either? as they're already failed binding.

Comment (michaelawyu) [inline pkg/controllers/rollout/controller.go:357]: Hi Ryan! Apologies if I misunderstand anything, but would this still get stuck if, say, all bindings are bound but failed to get applied (-> maxNumberToRemove <= 0) + maxUnavailable set to 0 with a non-zero maxSurge?

Comment (michaelawyu) [inline pkg/controllers/rollout/controller.go:357]: Hi Ryan! Apologies if I misunderstand anything, but would this still get stuck if, say, all bindings are bound but failed to get applied (-> maxNumberToRemove <= 0) + maxUnavailable set to 0 with a non-zero maxSurge?

Comment (ryanzhang-oss) [reply pkg/controllers/rollout/controller.go:357]: yes, but we also don't want to blast it too fast. This is kind of a middle ground.

Comment (ryanzhang-oss) [reply pkg/controllers/rollout/controller.go:357]: you mean when the "maxUnavailableNumber" == 0? Hmm.. that could be an issue.

**Review by zhiying-lin — APPROVED** [review]

**Pull request merged**